juice/src/net/layer.rs

```diff
@@ -69,3 +75,157 @@ impl From<SequentialBadInputOutputError> for LayerFromConfigError {
         LayerFromConfigError::Sequential(e)
     }
 }
+
+#[cfg(test)]
+pub mod testing {
+    use coaster::{Backend, ITensorDesc, Native, SharedTensor};
+
+    use crate::{
+        net::{Context, LearnableParamsLink, Network},
+        util::{native_backend, write_batch_sample},
+    };
+
+    #[derive(Debug)]
+    pub struct LayerOutput {
+        pub output: SharedTensor<f32>,
+    }
+
+    #[derive(Debug)]
+    pub struct LayerOutputAndGradients {
+        pub output: SharedTensor<f32>,
+        pub input_gradient: SharedTensor<f32>,
+        pub params_gradients: Vec<SharedTensor<f32>>,
+    }
+
+    // Set learnable params to given values. Panics if the overall sizes do not match.
+    pub fn set_params(params_link: &LearnableParamsLink, values: &[f32]) {
+        let mut params = params_link.borrow_mut();
+        assert_eq!(params.data.desc().size(), values.len());
+        let params_data = params.data.write_only(native_backend().device()).unwrap();
+        params_data.as_mut_slice::<f32>().copy_from_slice(values);
+    }
+
+    // Checks tensor equality. Convenience function for unit tests that allows to
+    // specify expected tensor as a 2D array. Example:
+    //   assert_tensor_eq(tensor, &[[1.0, -2.0]], &[[-3.0, 4.0]]);
+    pub fn assert_tensor_eq<Expected: AsRef<[ExpectedRow]>, ExpectedRow: AsRef<[f32]>>(
+        tensor: &SharedTensor<f32>,
+        expected: Expected,
+    ) {
+        let expected_rows = expected.as_ref().len();
+        let expected_cols = expected.as_ref()[0].as_ref().len();
+
+        // Make sure overall sizes match.
+        assert_eq!(expected_rows * expected_cols, tensor.desc().size());
+
+        let backend = native_backend();
+
+        let data = tensor.read(backend.device()).unwrap();
+        let slice = data.as_slice::<f32>();
+        for i in 0..expected_rows {
+            let mut equal = true;
+            for j in 0..expected_cols {
+                if (slice[i * expected_cols + j] - expected.as_ref()[i].as_ref()[j]).abs() > 0.00001 {
+                    equal = false;
+                    break;
+                }
+            }
+            if !equal {
+                panic!(
+                    "Row {} not equal: {:?} vs {:?}",
+                    i,
+                    &slice[i * expected_cols..(i + 1) * expected_cols],
+                    &expected.as_ref()[i].as_ref()
+                );
+            }
+        }
+    }
+
+    // Convenience function for layer unit tests. Example usage:
+    //   test_layer_compute(
+    //       LayerConfig::Relu,
+    //       &[[1.0, -2.0]], &[[-3.0, 4.0]],
+    //       &[[1.0, 0.0]], &[[0.0, 4.0]]);
```

Comment:
```suggestion
    //   ```
    //   test_layer_compute(
    //       LayerConfig::Relu,
    //       &[[1.0, -2.0]], &[[-3.0, 4.0]],
    //       &[[1.0, 0.0]], &[[0.0, 4.0]]);
    //   ```
```

Reply:
Yeah, stale comment. Updated.